aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/UploadOptions.java

```diff
@@ -0,0 +1,156 @@
+        this.bucket = builder.bucket;
+        this.metadata = builder.metadata;
+        this.cannedAcl = builder.cannedAcl;
+        this.listener = builder.listener;
+    }
+
+    public String getBucket() {
+        return this.bucket;
+    }
+
+    public ObjectMetadata getMetadata() {
+        return this.metadata;
+    }
+
+    public CannedAccessControlList getCannedAcl() {
+        return this.cannedAcl;
+    }
+
+    public TransferListener getTransferListener() {
+        return this.listener;
+    }
+
+    /**
+     * Builder class for UploadOptions
+     */
+    public static final class UploadOptionsBuilder {
+        private String bucket;
+        private ObjectMetadata metadata;
+        private CannedAccessControlList cannedAcl;
+        private TransferListener listener;
+
+        private UploadOptionsBuilder() {
+        }
+
+        /**
+         * Sets the bucket used for upload.
+         *
+         * @param bucket The bucket name
+         * @return builder
+         */
+        public UploadOptionsBuilder bucket(final String bucket) {
+            this.bucket = bucket;
+            return this;
+        }
+
+        /**
+         * Sets the metadata of the object to upload.
+         *
+         * @param metadata The S3 metadata to associate with this object
+         * @return builder
+         */
+        public UploadOptionsBuilder objectMetadata(final ObjectMetadata metadata) {
+            this.metadata = metadata;
+            return this;
+        }
+
+        /**
+         * Sets the canned ACL of the object to upload.
+         *
+         * @param cannedAcl The canned ACL to associate with this object
+         * @return builder
+         */
+        public UploadOptionsBuilder cannedAcl(final CannedAccessControlList cannedAcl) {
+            this.cannedAcl = cannedAcl;
+            return this;
+        }
+
+        /**
+         * Sets the listener to an upload task.
+         *
+         * @param listener The listener to attach to transfer observer
+         * @return builder
+         */
+        public UploadOptionsBuilder transferListener(final TransferListener listener) {
+            this.listener = listener;
+            return this;
+        }
+
+        /**
+         * Return the finally constructed UploadOptions.
+         * @return UploadOptions
+         */
+        public UploadOptions build() {
+            return new UploadOptions(this);
+        }
+    }
+
+    /**
+     * @return The builder object to construct a UploadOptions.
+     */
+    public static UploadOptionsBuilder builder() {
+        return new UploadOptionsBuilder();
+    }
+
+    @Override
+    public String toString() {
+        return "UploadOptions{" +
+                "bucket='" + bucket + '\'' +
+                ", metadata=" + metadata +
+                ", cannedAcl=" + cannedAcl +
+                ", listener=" + listener +
+                '}';
+    }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
For `amplify-android` there are a couple checkstyle rules that we follow to ensure a consistent style.  These aren't enforced on `aws-sdk-android`, mostly just because it would be too much effort to go fix all the style issues in the legacy code.  But, for new code like this, we should try to follow the same guidelines.   The ⌘N shortcut gives you a good starting point for `equals`, `hashcode` and `toString`, but some manual modifications are needed.  Here are a few things to address: 
 - While Java does allow you to skip the braces if you only have one statement after an if condition, we always include them anyway, with line breaks before and after, for better readability.
 - Avoid variable names with less than 3 characters.  Short variable names are difficult to search for, when trying to understand code.
 - `Objects.equals` and `Objects.hash` require API 19 and above.  To maintain compatibility with lower API versions, use `ObjectsCompat.equals` and `ObjectsCompat.hash` instead.

Reply:
Got it! That's a good practice for me to learn.😊